### ISSUE TITLE
ENH: add atol and rtol to np.unique to allow meaningful comparison of floating types

### ISSUE THREAD

**Issue by nfparham** [numpy/numpy#30072]
### Proposed new feature or change:

Similar to np.isclose, np.unique should accept an atol and rtol parameter to allow meaningful comparisons of floating values. Currently, two floating values that are off by a single machine epsilon are considered unique which is not necessarily useful behavior. Allowing both atol and rtol allows the caller to determine what threshold is meaningful for their use case. Default values of 0 for atol and rtol seem reasonable to keep the current behavior as the default.

**Comment by WarrenWeckesser** [comment]
A problem with attempting something like this is that the relation "is close to" is not transitive. That is, *`a` is close to `b`* and *`b` is close to `c`* does not imply that *`a` is close to `c`*.

This shows up in an example such as x = [0.0, 0.1, 0.2, 0.3] with, say,  `atol=0.125` and `rtol=0`.  What would `unique` return in that case?

**Comment by nfparham** [comment]
@WarrenWeckesser great point! Currently, unique returns the first observed entry as the unique index. Specifically, it states "The indices of the first occurrences of the unique values in the original array. Only provided if return_index is True.". This is not a universal definition but it was presumably adopted because it is useful. Similarly, the first entry could be used to compare the next n until the tolerance is exceeded and then that next entry could become the new point of comparison. I think it makes more sense to return the comparison value as-is rather than average the entries that were classified as duplicates since this would make it impossible to tell where the value came from in the original array and also allows a clear mapping between the value classified as unique and its index. Also, just like with np.isclose, the caller should know best as to what the right tolerance is to avoid cases like you described. However, I suppose this first n policy would eliminate the risk of such a chaining event. I'm not sure that warnings are very welcome to numpy, but I know I've seen them in the documentation and will cite one when I can find it - this would be deserving of such a warning.

So to address your example then,
* np.unique(x, atol=0.125, rtol=0) -> [0.0, 0.2]
* np.unique(x, atol=0.2, rtol=0) -> [0.0, 0.3]
* np.unique(x, atol=0.3, rtol=0.3) -> [0.0]
* np.unique(x, atol=1e-8, rtol=0) -> [0.0, 0.1, 0.2, 0.3]

And say x = [1, 1+1e-8, 2, 3], then it would return [1, 2, 3] for atol=1e-8 or [1+e-8, 2, 3] if the first two entries were swapped. Different, but not within the context the caller defined via the tolerance parameters. At least one desired post condition of the unique function is a reduction of the current array that omits duplicates. I’m not so sure in the floating case that an uber precise result is desired so much as a practically useful reduction.

**Comment by Tridib2510** [comment]
Is this issue resolved. Would love to contribute

**Comment by mattip** [comment]
Enhancements like this should hit the mailing list first for discussion, with a mention of how this is handled by the array api.

**Comment by seberg** [comment]
Yeah I am not sure I am convinced about in NumPy. Mostly because you need to argue that uaing a KDTree from scipy.spatial isn't the better/clearer alternative.

**Comment by charris** [comment]
I think what is desired -- rather than proposed -- is some sort of compact clustering. That seems out of scope.

**Comment by nfparham** [comment]
@charris per your point I do believe sklearn.cluster.DBSCAN(eps=1e-8, min_samples=1) is close enough to what I'm looking for

`x = np.array([1.0, 1+1e-8, 2.0, 3.0])`
`clusters = DBSCAN(eps=1e-8, min_samples=1).fit(x.reshape(-1, 1))`
`_, idx = np.unique(clusters.labels_, return_index=True)`
`x[idx]`
`array([1., 2., 3.])`


and just for sanity an out-of-order example

`x = np.array([1.0, 3.0, 2.0, 1+1e-8])`
`clusters = DBSCAN(eps=1e-8, min_samples=1).fit(x.reshape(-1, 1))`
`_, idx = np.unique(clusters.labels_, return_index=True)`
`x[idx]`
`array([1., 3., 2.])`

The tolerances don't always behave as expected but perhaps I need to mess around with the distance metric. Either way, seems like the clear and better option and I still get to make use of np.unique :)

Thanks @WarrenWeckesser @seberg @charris for your inputs! I think this is resolved

**Comment by WarrenWeckesser** [comment]
OK @nfparham, thanks for your interest in enhancing numpy!
